network-api/networkapi/wagtailpages/pagemodels/products.py

```diff
@@ -426,6 +503,87 @@ def get_context(self, request, *args, **kwargs):
+                        raise Http404("Product does not exist")
+
+                    # Save the new voting totals
+                    # TODO: Confirm with @pomax this is the intended behaviour we desire.
+                    product.current_vote_count = product.current_vote_count + value
+
+                    # Add the vote to the vote bin
+                    if not product.votes:
+                        # If there is no vote bin attached to this product yet, create one now.
+                        votes = ProductPageVotes()
+                        votes.save()
+                        product.votes = votes
+
+                    # Add the vote to the proper "vote bin"
+                    votes = product.votes.get_votes()
+                    if value <= 20:
+                        votes[0] = votes[0] + 1
+                    elif value <= 40:
+                        votes[1] = votes[1] + 1
+                    elif value <= 60:
+                        votes[2] = votes[2] + 1
+                    elif value <= 80:
+                        votes[3] = votes[3] + 1
+                    elif value <= 100:
+                        votes[4] = votes[4] + 1
+                    product.votes.set_votes(votes)
+
+                    # Don't save this as a revision with .save_revision() as to not spam the Audit log
+                    # And don't make this live with .publish(). The Page model will have the proper
+                    # data stored on it already, and the revision history won't be spammed by votes.
+                    product.save()
+                    return HttpResponse('Vote recorded', content_type='text/plain')
+                except ProductPage.DoesNotExist:
+                    return HttpResponse('Missing page', status=404, content_type='text/plain')
+                except ValidationError as ex:
+                    return HttpResponseNotAllowed(f'Payload validation failed: {ex}', content_type='text/plain')
+                except Error as ex:
+                    print(f'{ex.message} ({type(ex)})')
+                    return HttpResponseServerError('Internal Server Error', content_type='text/plain')
+        elif not self.votes:
+            # Double check a voting bin exists. It should always exist.
+            # TODO: Test the Product-to-ProductPage migration to ensure `votes` always exists.
+            # If all vote bins exist, we can safely remove this elif statement.
+            votes = ProductPageVotes()
+            votes.save()
+            self.votes = votes
+            self.save()
+
+        return super().serve(request, *args, **kwargs)
+
+    def save(self, *args, **kwargs):
+        # When a new ProductPage is created, ensure a vote bin always exists.
+        # We can use save() or a post-save Wagtail hook.
+        save = super().save(*args, **kwargs)
+        if not self.votes:
+            votes = ProductPageVotes()
+            votes.save()
+            self.votes = votes
```

Comment:
No not really. `serve()` is for serving the previews and published pages, `save()` is just for saving the page. I _don't think_ that `save()` gets called when a page is previewed though. Im going to run a test on this to see if we can safely remove these lines (566-573)

Reply:
Ok tested. Lines 566-573 are needed for previewing a page that hasnt been saved yet. In theory I thought this would be the case, in reality it's the case as well. 

But if the object is created when the page is viewed regarldess, we no longer need it in the `save()` method. I would prefer to have it in both areas for data consistency though in case we need to do a vote migration in the future and there are pages that havent been served that dont have `.votes` attached to them yet.

Reply:
Made it a method so both serve and save can get_or_create the votes.